src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/LanguageServiceHost.cs

```diff
@@ -265,10 +266,10 @@ public Task<bool> IsEnabledAsync(CancellationToken cancellationToken)
 
     public async Task WhenInitialized(CancellationToken token)
     {
-        await ValidateEnabledAsync(token);
-
         using (_joinableTaskCollection.Join())
         {
+            await ValidateEnabledAsync(token);
```

Comment:
This commit description for this change does not look correct. Join is about connecting the [JoinableTask created here](http://index/?leftProject=Microsoft.VisualStudio.ProjectSystem.Managed&leftSymbol=mmqz8avx2ktv&file=ProjectSystem%5CLanguageServices%5CWorkspace.cs&line=152) with the caller of this method. Anyone synchronously blocking on the UI thread on this method will happily let ValidateEnabledAsync switch to the UI thread if they are following JTF rules (ie calling JTF.Run).

Reply:
ie Join solely exists because there's not a traditional async connection between _firstPrimaryWorkspaceSet and that code I pointed to.

Reply:
> This commit description for this change does not look correct. Join is about connecting the [JoinableTask created here](http://index/?leftProject=Microsoft.VisualStudio.ProjectSystem.Managed&leftSymbol=mmqz8avx2ktv&file=ProjectSystem%5CLanguageServices%5CWorkspace.cs&line=152) with the caller of this method. Anyone synchronously blocking on the UI thread on this method will happily let ValidateEnabledAsync switch to the UI thread if they are following JTF rules (ie calling JTF.Run).

This JTC has all upstream data sources joined to it. The `Join` here allows dataflow to access the main thread, even if a caller of this method is on the main thread.

> the information whether it is running under command line is known upfront, the implementation could be further simplified and eliminate async - by getting this information using ICriticalPackageService

`ICriticalPackageService` is not available outside of CPS. I don't know a better way to initialize these variables early enough during solution load to ensure they're populated before we query them.

> moving ValidateEnabledAsync inside the collection join is unnecessary, because the AsyncLazy has its own JTF chain managed inside. it would not break things either.

If the join here was removed and the caller was on the main thread, and they trigger the lazy initialisation of the "is command line mode" state, which also require the main thread, but dataflow is running and is blocking the main thread, wouldn't it be able to deadlock?

Reply:
> ICriticalPackageService is not available outside of CPS. I don't know a better way to initialize these variables early enough during solution load to ensure they're populated before we query them.

Oh, I didn't realize that ICriticalPackageService is not public.
Another option would be to expose IVsShellServices.IsInCommandLineMode from CPS as public but that needs some additional thinking.

Reply:
>If the join here was removed and the caller was on the main thread, and they trigger the lazy initialisation of the "is command line mode" state, which also require the main thread, but dataflow is running and is blocking the main thread, wouldn't it be able to deadlock?

Join says "let any tasks added to the JoinableTaskCollection" access the main thread if the caller is blocking it. ValidateEnabledAsync does not add any tasks to the collection, and therefore putting it in the join has no affect.

Reply:
> ValidateEnabledAsync does not add any tasks to the collection

I thought tasks in the same collection are implicitly joined. The goal of having `ValidateEnabledAsync` inside the `Join` here is so that when it creates its own `JoinableTask` as part of the `AsyncLazy` (which uses the default JTF via MEF) that the `JoinableTask` instances associated with separate collections are considered as joined for the duration of that scope.

Reply:
Awaiting an AsyncLazy<T> implicitly joins the caller's context with the _Func<Task<T>>_ its executing, so if it needs UI thread, it will get it. All in all, there's probably no real side affect of doing what you are doing (other than allowing data to access UI thread earlier), but its unneeded and the lack of doing this will not cause a hang.